File: PageList.ts
export class Page {
  public previous: Page | null;
  public next: Page | null;
  public values: number[];
  constructor(m: number) {
    this.values = new Array(m);
    this.previous = null;
    this.next = null;
    this.values = [];
  }
}

export class PageList {
  public head: Page;
  private tail: Page;

  constructor(private p: number = 2) {
    const initialPage = new Page(this.p);
    this.head = initialPage;
    this.tail = initialPage;
  }

  public insert(value: number): number {
    let return_value; // returns 1 if a new Page is created, 0 otherwise

    // if tail is full, create a new Page in the list
    if (this.tail.values.length == this.p) {
      const newPage = new Page(this.p);
      this.tail.next = newPage;
      newPage.previous = this.tail;
      this.tail = newPage;

      return_value = 1;
    } else {
      return_value = 0;
    }

    this.tail.values.push(value);
    return return_value;
  }

  public print() {
    let response: string = "";
    let currentPage: Page | null = this.head;
    while (currentPage) {
      if (currentPage.next) {
        response += currentPage.values + " -> ";
      } else {
        response += currentPage.values;
      }
      currentPage = currentPage.next;
    }

    console.log(response);
  }
}

File: src/page_list.py
class Page:
    def __init__(self, m):
        self.m = m
        self.previous = None
        self.next = None
        self.values = []


class PageList:
    def __init__(self, p):
        self.p = p
        initial_page = Page(p)
        self.head = initial_page
        self.tail = initial_page
        self.page_count = 1

    def insert(self, value):
        return_value = 0

        if len(self.tail.values) == self.p:
            new_page = Page(self.p)
            self.tail.next = new_page
            new_page.previous = self.tail
            self.tail = new_page

            self.page_count += 1
            return_value = 1

        self.tail.values.append(value)
        return return_value

    def search(self, k):
        current_page = self.head
        access_counter = 0
        while current_page:
            access_counter += 1
            for value in current_page.values:
                if value == k:
                    return (value, access_counter)
            current_page = current_page.next

        return (-1, access_counter)

    def print(self):
        response = ""
        current_page = self.head
        while current_page:
            if current_page.next:
                response += str(current_page.values) + " -> "
            else:
                response += str(current_page.values)
            current_page = current_page.next
        print(response)
